File: rust_v2/examples/ws_ticker_and_prices.rs
//! WebSocket: best bid/ask ticker + price history with local candle state.
//!
//! Demonstrates:
//! - Ticker stream for real-time best bid/ask/mid updates
//! - `PriceHistoryState` for maintaining local candle data
//! - Handling the `PriceHistory` tagged enum (Snapshot, Update, Heartbeat)
//! - Graceful shutdown via Ctrl+C
//!
//! Set ORDERBOOK_ID in .env.
//!
//! ```bash
//! cargo run --example ws_ticker_and_prices --features native
//! ```

use futures_util::StreamExt;
use lightcone_sdk_v2::domain::price_history::wire::PriceHistory;
use lightcone_sdk_v2::prelude::*;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenvy::dotenv().ok();
    let client = LightconeClient::builder().build()?;

    let ob_id = OrderBookId::new(
        std::env::var("ORDERBOOK_ID").expect("Set ORDERBOOK_ID in .env"),
    );

    println!("Orderbook: {}", ob_id);
    println!("Press Ctrl+C to stop\n");

    let mut price_state = PriceHistoryState::new();
    let resolution = Resolution::Hour1;

    let mut ws = client.ws_native();
    ws.connect().await?;

    ws.send(MessageOut::subscribe_ticker(vec![ob_id.clone()]))?;
    ws.send(MessageOut::subscribe_price_history(ob_id.clone(), resolution))?;

    tokio::select! {
        _ = async {
            let mut stream = ws.events();
            while let Some(event) = stream.next().await {
                match event {
                    WsEvent::Connected => println!("[connected]"),

                    WsEvent::Message(Kind::Ticker(ticker)) => {
                        println!(
                            "[ticker] bid={} ask={} mid={}",
                            ticker.best_bid.map(|d| d.to_string()).unwrap_or_else(|| "—".into()),
                            ticker.best_ask.map(|d| d.to_string()).unwrap_or_else(|| "—".into()),
                            ticker.mid.map(|d| d.to_string()).unwrap_or_else(|| "—".into()),
                        );
                    }

                    WsEvent::Message(Kind::PriceHistory(ph)) => {
                        match ph {
                            PriceHistory::Snapshot(snap) => {
                                let prices: Vec<LineData> =
                                    snap.prices.into_iter().map(LineData::from).collect();
                                let count = prices.len();
                                price_state.apply_snapshot(
                                    ob_id.clone(),
                                    snap.resolution,
                                    prices,
                                );
                                println!(
                                    "[price snapshot] {} candles loaded ({})",
                                    count, snap.resolution
                                );
                                if let Some(data) = price_state.get(&ob_id, &snap.resolution) {
                                    if let Some(last) = data.last() {
                                        println!("  latest: ts={} value={}", last.time, last.value);
                                    }
                                }
                            }
                            PriceHistory::Update(upd) => {
                                let mid = upd.m.clone().unwrap_or_default();
                                let point = LineData {
                                    time: upd.t,
                                    value: mid.clone(),
                                };
                                price_state.apply_update(
                                    ob_id.clone(),
                                    upd.resolution,
                                    point,
                                );
                                let candle_count = price_state
                                    .get(&ob_id, &upd.resolution)
                                    .map(|d| d.len())
                                    .unwrap_or(0);
                                println!(
                                    "[price update] ts={} mid={} ({}) — {} candles total",
                                    upd.t, mid, upd.resolution, candle_count
                                );
                            }
                            PriceHistory::Heartbeat(hb) => {
                                println!(
                                    "[heartbeat] server_time={} last_processed={:?}",
                                    hb.server_time, hb.last_processed
                                );
                            }
                        }
                    }

                    WsEvent::Message(Kind::Pong(_)) => {}

                    WsEvent::Disconnected { code, reason } => {
                        println!("[disconnected] code={:?} reason={}", code, reason);
                    }

                    WsEvent::MaxReconnectReached => {
                        eprintln!("[fatal] max reconnect attempts reached");
                        break;
                    }

                    WsEvent::Error(e) => eprintln!("[error] {}", e),
                    WsEvent::Message(Kind::Error(e)) => eprintln!("[server error] {}", e),
                    _ => {}
                }
            }
        } => {}
        _ = tokio::signal::ctrl_c() => {
            println!("\nShutting down...");
        }
    }

    ws.disconnect().await?;

    if let Some(data) = price_state.get(&ob_id, &resolution) {
        println!("\nFinal candle count ({}): {}", resolution, data.len());
    }

    Ok(())
}

File: python/examples/ws_ticker_and_prices.py
"""Best bid/ask ticker + price history line data with PriceHistoryState."""

import asyncio

from common import rest_client, market_and_orderbook
from lightcone_sdk.ws import WsEventType, MessageInType
from lightcone_sdk.ws.subscriptions import TickerParams, PriceHistoryParams
from lightcone_sdk.domain.price_history.state import PriceHistoryState
from lightcone_sdk.domain.price_history import LineData
from lightcone_sdk.domain.price_history.wire import (
    PriceHistorySnapshot,
    PriceHistoryUpdate,
    PriceHistoryHeartbeat,
)
from lightcone_sdk.shared.types import Resolution


async def main():
    client = rest_client()
    _, orderbook = await market_and_orderbook(client)
    orderbook_id = orderbook.orderbook_id
    resolution = Resolution.ONE_MINUTE
    resolution_str = resolution.as_str()

    # State tracker
    history = PriceHistoryState()

    # Connect WebSocket
    ws = client.ws()
    await ws.connect()
    print("connected")

    # Subscribe to ticker and price history
    await ws.subscribe(TickerParams(orderbook_ids=[orderbook_id]))
    await ws.subscribe(PriceHistoryParams(
        orderbook_id=orderbook_id,
        resolution=resolution.as_str(),
        include_ohlcv=False,
    ))

    event_count = 0
    max_events = 4
    done = asyncio.Event()

    def on_event(event):
        nonlocal event_count
        if event.type == WsEventType.MESSAGE and event.message:
            msg = event.message

            if msg.type == MessageInType.TICKER.value:
                ticker = msg.data
                print(
                    f"ticker: bid={ticker.best_bid} "
                    f"ask={ticker.best_ask} "
                    f"mid={ticker.mid_price}"
                )
                event_count += 1

            elif msg.type == MessageInType.PRICE_HISTORY.value:
                data = msg.data

                if isinstance(data, PriceHistorySnapshot):
                    prices = [
                        LineData(time=c.t, value=c.m or "0")
                        for c in data.candles
                    ]
                    history.apply_snapshot(
                        data.orderbook_id, data.resolution, prices
                    )
                    candles = history.get(orderbook_id, resolution_str)
                    print(f"price snapshot: {len(candles)} candle(s)")
                    event_count += 1

                elif isinstance(data, PriceHistoryUpdate):
                    if data.candle:
                        point = LineData(
                            time=data.candle.t, value=data.candle.m or "0"
                        )
                        history.apply_update(
                            data.orderbook_id, data.resolution, point
                        )
                        candles = history.get(orderbook_id, resolution_str)
                        latest = candles[-1] if candles else None
                        print(f"latest candle: {latest}")
                        event_count += 1

                elif isinstance(data, PriceHistoryHeartbeat):
                    print(f"heartbeat: {data.server_time}")

        elif event.type == WsEventType.ERROR:
            print(f"ws error: {event.error}")

        if event_count >= max_events:
            done.set()

    ws.on(on_event)

    try:
        await asyncio.wait_for(done.wait(), timeout=30)
    except asyncio.TimeoutError:
        pass

    await ws.disconnect()
    await client.close()


asyncio.run(main())
